File: backend/app/domains/collections/schemas/collections.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

CollectionAccessPolicy = Literal["org_wide", "admin_only", "selected_roles", "selected_members"]

GranteeType = Literal["role", "member"]

RuleField = Literal[
    "file_type",
    "language",
    "status",
    "ingestion_source",
    "trust_status",
    "uploaded_by_user_id",
    "tags",
]

RuleOperator = Literal["eq", "neq", "in", "not_in", "contains", "not_contains"]

RuleLogic = Literal["and", "or"]


class RuleCondition(BaseModel):
    field: RuleField
    operator: RuleOperator
    value: Any = Field(..., description="String for scalar operators; list for 'in'/'not_in'")


class DynamicRuleSet(BaseModel):
    logic: RuleLogic = "and"
    conditions: list[RuleCondition] = Field(..., min_length=1, max_length=20)


class CollectionListItemResponse(BaseModel):
    collection_id: str
    name: str
    description: str | None
    owner_id: str
    owner_email: str | None
    document_count: int
    indexed_count: int
    access_policy: CollectionAccessPolicy
    is_dynamic: bool
    last_rule_evaluated_at: datetime | None
    created_at: datetime
    updated_at: datetime


class CollectionDetailResponse(CollectionListItemResponse):
    created_by_email: str | None
    rule_schema: DynamicRuleSet | None


class CollectionListResponse(BaseModel):
    items: list[CollectionListItemResponse]
    total: int


class CreateCollectionRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    access_policy: CollectionAccessPolicy = "org_wide"
    is_dynamic: bool = False
    rule_schema: DynamicRuleSet | None = None


class UpdateCollectionRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = None
    access_policy: CollectionAccessPolicy | None = None


class DeleteCollectionResponse(BaseModel):
    collection_id: str
    archived: bool


class CollectionDocumentItem(BaseModel):
    document_id: str
    filename: str
    file_type: str
    status: str
    updated_at: datetime


class CollectionDocumentsResponse(BaseModel):
    items: list[CollectionDocumentItem]
    total: int


class AddDocumentToCollectionRequest(BaseModel):
    document_id: str


class AddDocumentToCollectionResponse(BaseModel):
    collection_id: str
    document_id: str


class DocumentCollectionsResponse(BaseModel):
    items: list[CollectionListItemResponse]


class SetDocumentCollectionsRequest(BaseModel):
    collection_ids: list[str]


# ── Access policy management ───────────────────────────────────────────────────


class CollectionAccessGrantItem(BaseModel):
    grantee_type: GranteeType
    grantee_value: str  # role name (e.g. "member") or user_id string


class CollectionPolicyResponse(BaseModel):
    collection_id: str
    access_policy: CollectionAccessPolicy
    grants: list[CollectionAccessGrantItem]


class UpdateCollectionPolicyRequest(BaseModel):
    access_policy: CollectionAccessPolicy
    grants: list[CollectionAccessGrantItem] = Field(default_factory=list)


# ── Dynamic rule management ────────────────────────────────────────────────────


class SetCollectionRulesRequest(BaseModel):
    rule_schema: DynamicRuleSet


class CollectionRulesResponse(BaseModel):
    collection_id: str
    is_dynamic: bool
    rule_schema: DynamicRuleSet | None
    last_rule_evaluated_at: datetime | None
    matched_count: int


class PreviewRulesRequest(BaseModel):
    rule_schema: DynamicRuleSet
    limit: int = Field(default=20, ge=1, le=100)


class PreviewRulesDocumentItem(BaseModel):
    document_id: str
    filename: str
    file_type: str
    language: str | None
    status: str
    trust_status: str | None
    tags: str | None
    ingestion_source: str | None


class PreviewRulesResponse(BaseModel):
    total: int
    items: list[PreviewRulesDocumentItem]


class RefreshRulesResponse(BaseModel):
    collection_id: str
    matched_count: int
    last_rule_evaluated_at: datetime | None

File: frontend/src/lib/api/collections.ts
import { apiRequest } from "@/lib/api/request";
import type { DocumentListItemResponse } from "@/lib/api/documents";

export type CollectionAccessPolicy =
  | "org_wide"
  | "admin_only"
  | "selected_roles"
  | "selected_members";

export type GranteeType = "role" | "member";

export type RuleField =
  | "file_type"
  | "language"
  | "status"
  | "ingestion_source"
  | "trust_status"
  | "uploaded_by_user_id"
  | "tags";

export type RuleOperator =
  | "eq"
  | "neq"
  | "in"
  | "not_in"
  | "contains"
  | "not_contains";

export type RuleLogic = "and" | "or";

export type RuleCondition = {
  field: RuleField;
  operator: RuleOperator;
  value: string | string[];
};

export type DynamicRuleSet = {
  logic: RuleLogic;
  conditions: RuleCondition[];
};

export type CollectionListItemResponse = {
  collection_id: string;
  name: string;
  description: string | null;
  owner_id: string;
  owner_email: string | null;
  document_count: number;
  indexed_count: number;
  access_policy: CollectionAccessPolicy;
  is_dynamic: boolean;
  last_rule_evaluated_at: string | null;
  review_status?:
    | "current"
    | "trusted"
    | "needs_review"
    | "stale"
    | "expired"
    | "archived";
  review_owner_id?: string | null;
  review_due_date?: string | null;
  expiry_date?: string | null;
  trust_level?: string | null;
  created_at: string;
  updated_at: string;
};

export type CollectionDetailResponse = CollectionListItemResponse & {
  created_by_email: string | null;
  rule_schema: DynamicRuleSet | null;
};

export type CollectionListResponse = {
  items: CollectionListItemResponse[];
  total: number;
  freshness?: CollectionListItemResponse["review_status"] | null;
};

export type CreateCollectionRequest = {
  name: string;
  description?: string | null;
  access_policy?: CollectionAccessPolicy;
  is_dynamic?: boolean;
  rule_schema?: DynamicRuleSet | null;
};

export type UpdateCollectionRequest = {
  name?: string;
  description?: string | null;
  access_policy?: CollectionAccessPolicy;
  review_status?:
    | "current"
    | "trusted"
    | "needs_review"
    | "stale"
    | "expired"
    | "archived"
    | null;
  review_owner_id?: string | null;
  review_due_date?: string | null;
  expiry_date?: string | null;
  trust_level?: string | null;
};

export type DeleteCollectionResponse = {
  collection_id: string;
  archived: boolean;
};

export type CollectionDocumentsResponse = {
  items: DocumentListItemResponse[];
  total: number;
};

export type AddDocumentToCollectionResponse = {
  collection_id: string;
  document_id: string;
};

export type DocumentCollectionsResponse = {
  items: CollectionListItemResponse[];
};

export type ListCollectionsOptions = {
  limit?: number;
  offset?: number;
  name_query?: string;
  freshness?:
    | "current"
    | "trusted"
    | "needs_review"
    | "stale"
    | "expired"
    | "archived";
};

// ── Access policy types ────────────────────────────────────────────────────────

export type CollectionAccessGrant = {
  grantee_type: GranteeType;
  grantee_value: string; // role name or user_id
};

export type CollectionPolicyResponse = {
  collection_id: string;
  access_policy: CollectionAccessPolicy;
  grants: CollectionAccessGrant[];
};

export type UpdateCollectionPolicyRequest = {
  access_policy: CollectionAccessPolicy;
  grants?: CollectionAccessGrant[];
};

// ── Collection CRUD ────────────────────────────────────────────────────────────

export async function listCollections(
  options: ListCollectionsOptions = {},
): Promise<CollectionListResponse> {
  return apiRequest<CollectionListResponse>("/collections", {
    query: {
      limit: options.limit,
      offset: options.offset,
      name_query: options.name_query || undefined,
      freshness: options.freshness,
    },
  });
}

export async function getCollection(
  collectionId: string,
): Promise<CollectionDetailResponse> {
  return apiRequest<CollectionDetailResponse>(
    `/collections/${encodeURIComponent(collectionId)}`,
  );
}

export async function createCollection(
  payload: CreateCollectionRequest,
): Promise<CollectionDetailResponse> {
  return apiRequest<CollectionDetailResponse>("/collections", {
    method: "POST",
    json: payload,
  });
}

export async function updateCollection(
  collectionId: string,
  payload: UpdateCollectionRequest,
): Promise<CollectionDetailResponse> {
  return apiRequest<CollectionDetailResponse>(
    `/collections/${encodeURIComponent(collectionId)}`,
    {
      method: "PATCH",
      json: payload,
    },
  );
}

export async function deleteCollection(
  collectionId: string,
): Promise<DeleteCollectionResponse> {
  return apiRequest<DeleteCollectionResponse>(
    `/collections/${encodeURIComponent(collectionId)}`,
    { method: "DELETE" },
  );
}

// ── Access policy management ───────────────────────────────────────────────────

export async function getCollectionPolicy(
  collectionId: string,
): Promise<CollectionPolicyResponse> {
  return apiRequest<CollectionPolicyResponse>(
    `/collections/${encodeURIComponent(collectionId)}/access-policy`,
  );
}

export async function updateCollectionPolicy(
  collectionId: string,
  payload: UpdateCollectionPolicyRequest,
): Promise<CollectionPolicyResponse> {
  return apiRequest<CollectionPolicyResponse>(
    `/collections/${encodeURIComponent(collectionId)}/access-policy`,
    {
      method: "PUT",
      json: payload,
    },
  );
}

// ── Collection documents ───────────────────────────────────────────────────────

export async function listCollectionDocuments(
  collectionId: string,
  options: {
    limit?: number;
    offset?: number;
    freshness?:
      | "current"
      | "trusted"
      | "needs_review"
      | "stale"
      | "expired"
      | "archived";
  } = {},
): Promise<CollectionDocumentsResponse> {
  return apiRequest<CollectionDocumentsResponse>(
    `/collections/${encodeURIComponent(collectionId)}/documents`,
    {
      query: {
        limit: options.limit,
        offset: options.offset,
        freshness: options.freshness,
      },
    },
  );
}

export async function addDocumentToCollection(
  collectionId: string,
  documentId: string,
): Promise<AddDocumentToCollectionResponse> {
  return apiRequest<AddDocumentToCollectionResponse>(
    `/collections/${encodeURIComponent(collectionId)}/documents`,
    {
      method: "POST",
      json: { document_id: documentId },
    },
  );
}

export async function removeDocumentFromCollection(
  collectionId: string,
  documentId: string,
): Promise<void> {
  await apiRequest<unknown>(
    `/collections/${encodeURIComponent(collectionId)}/documents/${encodeURIComponent(documentId)}`,
    { method: "DELETE" },
  );
}

export async function getDocumentCollections(
  documentId: string,
): Promise<DocumentCollectionsResponse> {
  return apiRequest<DocumentCollectionsResponse>(
    `/documents/${encodeURIComponent(documentId)}/collections`,
  );
}

export async function setDocumentCollections(
  documentId: string,
  collectionIds: string[],
): Promise<DocumentCollectionsResponse> {
  return apiRequest<DocumentCollectionsResponse>(
    `/documents/${encodeURIComponent(documentId)}/collections`,
    {
      method: "PUT",
      json: { collection_ids: collectionIds },
    },
  );
}

// ── Dynamic rules ──────────────────────────────────────────────────────────────

export type CollectionRulesResponse = {
  collection_id: string;
  is_dynamic: boolean;
  rule_schema: DynamicRuleSet | null;
  last_rule_evaluated_at: string | null;
  matched_count: number;
};

export type PreviewRulesDocumentItem = {
  document_id: string;
  filename: string;
  file_type: string;
  language: string | null;
  status: string;
  trust_status: string | null;
  tags: string | null;
  ingestion_source: string | null;
};

export type PreviewRulesResponse = {
  total: number;
  items: PreviewRulesDocumentItem[];
};

export type RefreshRulesResponse = {
  collection_id: string;
  matched_count: number;
  last_rule_evaluated_at: string | null;
};

export async function setCollectionRules(
  collectionId: string,
  ruleSchema: DynamicRuleSet,
): Promise<CollectionRulesResponse> {
  return apiRequest<CollectionRulesResponse>(
    `/collections/${encodeURIComponent(collectionId)}/rules`,
    {
      method: "PUT",
      json: { rule_schema: ruleSchema },
    },
  );
}

export async function previewCollectionRules(
  collectionId: string,
  ruleSchema: DynamicRuleSet,
  limit = 20,
): Promise<PreviewRulesResponse> {
  return apiRequest<PreviewRulesResponse>(
    `/collections/${encodeURIComponent(collectionId)}/rules/preview`,
    {
      method: "POST",
      json: { rule_schema: ruleSchema, limit },
    },
  );
}

export async function refreshCollectionRules(
  collectionId: string,
): Promise<RefreshRulesResponse> {
  return apiRequest<RefreshRulesResponse>(
    `/collections/${encodeURIComponent(collectionId)}/rules/refresh`,
    { method: "POST" },
  );
}
